beacon-chain/sync/data_columns_sampling.go

```diff
@@ -273,132 +294,188 @@ func (s *Service) incrementalDAS(
 			return false, nil, errors.New("retrieved more columns than requested")
 		}
 
-		// Some columns are missing, we need to extend the sample size.
-		missingColumnsCount += columnsToSampleCount - retrievedSampleCount
-
-		firstColumnToSample = extendedSampleCount
+		// missing columns, extend the samples.
+		allowedFailures += columnsToSampleCount - retrievedSampleCount
 		oldExtendedSampleCount := extendedSampleCount
-		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, missingColumnsCount)
+		firstColumnToSample = extendedSampleCount
+		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, allowedFailures)
 
 		log.WithFields(logrus.Fields{
 			"root":                fmt.Sprintf("%#x", root),
 			"round":               round,
-			"missingColumnsCount": missingColumnsCount,
-			"currentSampleCount":  oldExtendedSampleCount,
-			"nextSampleCount":     extendedSampleCount,
+			"missingColumnsCount": allowedFailures,
+			"currentSampleIndex":  oldExtendedSampleCount,
+			"nextSampleIndex":     extendedSampleCount,
 		}).Debug("Some columns are still missing after sampling this round.")
 	}
 }
 
-// DataColumnSamplingRoutine runs incremental DAS on block when received.
-func (s *Service) DataColumnSamplingRoutine(ctx context.Context) {
-	// Get the custody subnets count.
-	custodySubnetsCount := peerdas.CustodySubnetCount()
-
-	// Create a subscription to the state feed.
-	stateChannel := make(chan *feed.Event, 1)
-	stateSub := s.cfg.stateNotifier.StateFeed().Subscribe(stateChannel)
+func (d *dataColumnSampler1D) sampleDataColumns(
+	ctx context.Context,
+	root [fieldparams.RootLength]byte,
+	columns []uint64,
+) (map[uint64]bool, error) {
+	// distribute samples to peer
+	peerToColumns, err := d.distributeSamplesToPeer(columns)
+	if err != nil {
+		log.WithError(err).Error("Failed to distribute samples to peers")
+		return nil, errors.Wrap(err, "failed distributing samples to peer")
+	}
 
-	// Unsubscribe from the state feed when the function returns.
-	defer stateSub.Unsubscribe()
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	res := make(map[uint64]bool)
+	sampleFromPeer := func(pid peer.ID, cols map[uint64]bool) {
+		defer wg.Done()
+		retrieved := d.sampleDataColumnsFromPeer(ctx, pid, root, cols)
 
-	// Retrieve the number of columns.
-	columnsCount := params.BeaconConfig().NumberOfColumns
+		mu.Lock()
+		for col := range retrieved {
+			res[col] = true
+		}
+		mu.Unlock()
+	}
 
-	// Retrieve all columns we custody.
-	custodyColumns, err := peerdas.CustodyColumns(s.cfg.p2p.NodeID(), custodySubnetsCount)
-	if err != nil {
-		log.WithError(err).Error("Failed to get custody columns")
-		return
+	// sample from peers in parallel
+	for pid, cols := range peerToColumns {
+		wg.Add(1)
+		go sampleFromPeer(pid, cols)
 	}
```

Comment:
In the worst case, with `n` the number of samples, we can query `n` different peers in parallel at the same time.
I'm not totally sure of the implication regarding networking of this fact, especially for our Windows users, where Windows tends to consider Prysm as a malware in this kind of situations.

@nisdas do you have any take on that? Should we limit the number of parallel queries?

Reply:
My 2 cents here are for average cases, the parallelization won't be too bad, 16 samples initially and incremental DAS won't dramatically add more sample queries.

But if we want to be absolutely safe, we could agree on a number and limit concurrency here